### ISSUE TITLE
rustdoc removes Try from <Self as Try>::Residual in std::ops::FromResidual

### ISSUE THREAD

**Issue by veber-alex** [rust-lang/rust#85454]
In the docs for the new FromResidual trait here:
https://doc.rust-lang.org/nightly/std/ops/trait.FromResidual.html

The trait is shown as:
```rust
pub trait FromResidual<R = Self::Residual> {
    fn from_residual(residual: R) -> Self;
}
```
which is not valid and will not compile, in the source it looks like this:
```rust
pub trait FromResidual<R = <Self as Try>::Residual> {
    fn from_residual(residual: R) -> Self;
}
```

**Comment by scottmcm** [comment]
Hypothesis: this a a weird case in the associated type normalization, since in just about any other position this would only be allowed because there'd be a `where Self: Try` in scope, but there's isn't here.

(It's a very unusual case, and also broke R-A: https://github.com/rust-analyzer/rust-analyzer/pull/8692)

**Comment by Stupremee** [comment]
I researched a bit into this issue. The problem is that rustdoc never renders a `<Self as X>::Y` cast (where `Self` is the type to be cast) because of this check:
https://github.com/rust-lang/rust/blob/9f8012e3aa8d5b087100ecd7b109ed3dd2859cca/src/librustdoc/html/format.rs#L842

However, it's a bit harder to fix this because simply removing this check would lead to false-positive where `Self` is cast to the type of itself (`<Self as Deref>::Target` in a `Deref` impl for example), which is wrong too.

**Comment by Stupremee** [comment]
This needs to be re-opened because it does not work if the trait is re-exported. 
Compare https://doc.rust-lang.org/nightly/std/ops/trait.FromResidual.html with https://doc.rust-lang.org/nightly/core/ops/trait.FromResidual.html

**Comment by cynecx** [comment]
@Stupremee https://github.com/rust-lang/rust/issues/84579 might be related.

**Comment by Stupremee** [comment]
Cross Crate re-exports are *really weird* in rustdoc. They already caused me so much pain

**Comment by cynecx** [comment]
@Stupremee https://github.com/rust-lang/rust/issues/84579#issuecomment-848254819 might be interesting.
